### PR TITLE
[BEAM-7268] make sorter extension Hadoop-free

### PR DESCRIPTION
**Pull request by nevillelyh**
Right now the Java sorter extension depends on Hadoop SequenceFile for external sort. It'll be nice to re-implement it without the dependency to avoid conflicts.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by nevillelyh** [comment]
R: @kanterov

**Comment by kanterov** [comment]
Great to reduce dependencies. If dependency conflicts are the only concern, an alternative could be shading Hadoop dependencies. It's hard to say without having benchmarks if this code is going to perform better or not. I would suggest having it as an alternative implementation for sorting and keep the previous one.

If I'm not mistaken, your work for `SortValues` is needed to express `GroupByKeyAndSortValues` as `GroupByKey`+SortValues`. There is a thread on a mailing list on adding `GroupByKeyAndSortValues` as a primitive transform in Beam [gkb-sort-values]. I would rather go that way, runners such as Dataflow, Spark, and Flink would override it with efficient implementation because they have a concept of secondary sorting, that would be way more efficient than anything else.

What do you think about rather investing in having `GroupByKeyAndSortValues`, because, in the end, that's what we need? I might be better to move this discussion into the mailing list.

cc @kennknowles @reuvenlax 

[gbk-sort-values]: https://lists.apache.org/thread.html/313934c8543ae84d541653050a1bc77b5550b4a8262afd51e5695365@%3Cdev.beam.apache.org%3E

**Comment by kennknowles** [comment]
Seems like a fine idea to me. It is a good point that you could just add it as a parallel extension. Or do fancy build tricks. But probably it is just easy and clean to have a separate library. It will be a little while before I can give this a good review, and I think @reuvenlax is the better choice anyhow.

**Comment by nevillelyh** [comment]
Agree with both. It's still worth having a "lean" option though. I'll work on some benchmark & adding it as a parallel implement, e.g. `{Native,Hadoop}ExternalSorter`?

**Comment by nevillelyh** [comment]
Did some basic benchmark, with 5 million UUID KV pairs (~360MB raw bytes), 64MB memory, Hadoop vs native impl is ~51.4s vs ~17.9s

**Comment by nevillelyh** [comment]
I've made the change backwards compatible and added the native impl as an option. @kanterov @reuvenlax PTAL?

**Comment by kanterov** [comment]
I don't have time to review before the 10th of June.

**Comment by lukecwik** [comment]
@kanterov Were you planning to take a look at this?

File: sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/NativeFileSorter.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.Iterators;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.primitives.UnsignedBytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * External Sorter based on <a
+ * href="https://github.com/lemire/externalsortinginjava">lemire/externalsortinginjava</a>.
+ */
+public class NativeFileSorter {
```
Comment (kanterov) [inline sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/NativeFileSorter.java]: nit: probably should be package-private

Comment (nevillelyh) [reply sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/NativeFileSorter.java]: Fixed.

**Comment by kanterov** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by kanterov** [comment]
сс @lemire

**Comment by lemire** [comment]
@kanterov  Thanks for pinging me. How can I help?

**Comment by kanterov** [comment]
@lemire thanks, no help is needed, just thought that you might want to know

**Comment by lemire** [comment]
Indeed. Glad to see our code being used.

**Pull request merged**